### ISSUE TITLE
Proposal: add a DcaReader to access DCA data in a structured way

### ISSUE THREAD

**Issue by SeverinGloeckle** [contao/contao#3687]
I would like to discuss a possibility to standardize and ease the reading of configuration data from a DCA.

While the definition of DCAs via `$GLOBALS['TL_DCA']` cannot be changed at the moment, it would be nice if one could read values from the global array in a more structured way.

## Feature proposal: `DcaReader`
I'd suggest something like a `DcaReader`, which provides a method to get the value of any node within a DCA using dot notation:

````php
$mode = $dcaReader->get('list.sorting.mode');
```` 

**Caveats of dot notation**
While the dot notation is very convenient, there can be problems with nodes that contain a dot in their key. This could be handled explicitly in the code (and should probably be deprecated). Symfony's [PropertyAcccess](https://symfony.com/doc/current/components/property_access.html) therefore uses an array-like index notation, but in my opinion a dot notation is way easier to write and read.

**Utility methods**
Additionally, various utility methods could provide ways to query common data from the DCA:
- `field($field)` to get a field
- `field($field, 'key')` to get a subset of a field
- `isRelatedField($field)` to check whether a given field has a relation
- `isClosed()` for `config.closed`
- `getOperations` for `list.operations`
- ...

## Schema assets for DCAs
Even better than utility methods would be a definition of various subset schemas for a DCA, similar to Doctrine's [Schema Assets](https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/schema-representation.html#schema-assets). 
As a quick thought, the following types come to mind:
- `DCA\Schema\Config`
- `DCA\Schema\CallbackCollection`
- `DCA\Schema\Callback`
- `DCA\Schema\Sorting`
- `DCA\Schema\Label`
- `DCA\Schema\GlobalOperation`
- `DCA\Schema\Operation`
- `DCA\Schema\Palette`
- `DCA\Schema\Field`
- `DCA\Schema\SQL` (which probably would extend Doctrine's `Column` asset)

The various assets then can provide dedicated utility methods.

**Custom DCA fields**
Since the various schema classes can only handle standard DCA fields, custom DCA fields would have to be accessed via some `get()` method that reads from the data source.

## Exemplary pseudo-code
An implementation could be used similarly to the following:

````php
// Get a reader from an injected factory or from a service locator / collection
$dcaReader = $this->dcaReaderCollection->get('tl_page');
````

````php
// Current:
$this->treeView = \in_array($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null, array(5, 6));

// With reader:
$this->treeView = \in_array($dcaReader->get('list.sorting.mode'), array(5,6));

// Or with some schema classes:
$this->treeView = $dcaReader->sorting->isTreeView();
````

````php
// Current:
if (\is_array($GLOBALS['TL_DCA'][$dc->table]['fields'][$strField]['load_callback'] ?? null))
{
  foreach ($GLOBALS['TL_DCA'][$dc->table]['fields'][$strField]['load_callback'] as $callback)
  {
    if (\is_array($callback))
    {
      $this->import($callback[0]);
      $varValue = $this->{$callback[0]}->{$callback[1]}($varValue, $dc);
    }
    elseif (\is_callable($callback))
    {
      $varValue = $callback($varValue, $dc);
    }
  }
}

// With reader:
if (\is_array($dcaReader->field($strField, 'load_callback')))
{
  foreach ($dcaReader->field($strField, 'load_callback') as $callback)
  {
    if (\is_array($callback))
    {
      $this->import($callback[0]);
      $varValue = $this->{$callback[0]}->{$callback[1]}($varValue, $dc);
    }
    elseif (\is_callable($callback))
    {
      $varValue = $callback($varValue, $dc);
    }
  }
}

// Or with some schema classes
$varValue = $dcaReader->field($strField)->getCallback('load')->call($varValue, $dc);
````

The benefits are pretty obvious:
- Easier access to DCA data without lots of `isset` and boilerplate code.
- A common interface to access DCA data makes other sources for DCA configuration possible in the future (yaml, custom code, API, ...).
- Code that is reading data from a DCA can be tested more easily since the reader is mockable.
- An IDE can provide autocompletion for DCA data, adding some kind of in-code documentation.

As each reader is just an alternative way to read data from `$GLOBALS['TL_DCA']`, it's possible to use the reader step-by-step in throughout the application.

**Possibility for DCA objects**
While the first step is just a DCA reader, it should be possible to create a real DCA object on top of the reader in the future which also allows to set DCA data (writing it to the global array in the background).

## Note on legacy code
For legacy code without proper dependency injection, one could even think of a service locator to avoid using a factory to get a reader:

````php
$label = System::dca($table)->field('headline', 'label');
````

-----

I'm interested in opinions on this feature. If it sounds like a good addition, I can work on a first draft.

**Comment by ausi** [comment]
> While the dot notation is very convenient, there can be problems with nodes that contain a dot in their key.

We solved this problem already in our translator decoration: https://github.com/contao/contao/blob/222659106d851a50f243e29db0995143fcf2d51f/core-bundle/tests/Translation/TranslatorTest.php#L104-L112

**Comment by Toflar** [comment]
I like the idea and I would love to see a first draft (can be very rough) just so we can see if the concept would even work.
One thing I would like to get the feedback of the other @contao/developers: Maybe it's time to get rid of `DCA`. DCA stands for `DataContainerArray` and it's not really an array anymore. Or well, it shouldn't if we start going into a different direction here. I think at the end of the day we're still talking "data container" because that's just what we name it but probably it's `DataContainerConfig` rather than array?

**Comment by aschempp** [comment]
I totally understand the idea and it would theoretically be a nice addition. I think we discussed and/or tried stuff like this before (similar to using YAML for DCA), but it basically boils down to first discuss how the future of DCA/C might look like. If we agree it will stay like this, a reader might be useful. But we also discussed using a Symfony ConfigTree or similar for it, and I'm not sure the reader would then prevent us from doing any improvements...

**Comment by Toflar** [comment]
Agreed. I think we can agree that we'd love to have a solution that works similar to the Symfony bundle config. At the end of the day, the DCA is no different. It's just a huge configuration array.
The nice thing about the regular config tree is that we could support all formats the Symfony `Config` component supports too. We'd need to have a compatibility layer for the global array style syntax we have now.

**Comment by SeverinGloeckle** [comment]
> I think we can agree that we'd love to have a solution that works similar to the Symfony bundle config. [...]
> We'd need to have a compatibility layer for the global array style syntax we have now.

That's pretty much what I'm working on after @m-vo suggested the idea of using Symfony's config bundle to me.

My current draft flows like this:
- The structure of a general DCA is defined in a `DcaConfiguration` which implements Symfony's `ConfigurationInterface`.
- The configuration tree extends the tree of Symfony's config bundle with so called schema nodes which can map data to a schema class (e.g. `Config`, `CallbackCollection` or `Operation`).
- There is a new service `DcaFactory` that creates a `Dca` object for a given key (e.g. `tl_content`).
- Internally, the factory resolves a driver for the key that provides a simple array of data.
- The only driver at first will be a `GlobalArrayDriver` that reads the data from `$GLOBALS['TL_DCA']`, but there could be others in the future.
- The factory processes the data via Symfony's config processor using the`DcaConfiguration` tree.
- The driver may attach observers to the data so the configuration may get reprocessed in case the source data changes (e.g. when modifying `$GLOBALS['TL_DCA']` directly).
- The generated `Dca` object wraps the processed configuration array and provides a fluent interface to work with the data (e.g. `$dca->config()->isClosed()`).
- Most utility functions are implemented via the schema classes (e.g. `config()->getCallback('onload')->call()`).
- Any/most extra keys from the original data will still be available in the Dca (e.g. `$dca->list()->get('custom.mode.flag')`).


Additional notes:
- All nodes of the `Dca` object are based on `ArrayObject` and thus are dumpable to YAML. 
- It's possible to define a strict structure for DCAs via the configuration tree. For example, any custom keys under `config` could be disallowed.

**About the name DCA**
In my opinion, the name _data container array_ may still be valid since it will still be based on a configuration **array**, which is used to configure data for a **data container**, hence data container array.

**Current state**
I have the basic concept up and running, but am still working on:
- Drafting/testing more schema nodes to detect possible problems (callbacks still got my head spinning).
- Providing an easy way to use the `Dca` object in tests.
- Validating which backwards compatibility features are required.
- Adding a possibility to customize the DCA configuration tree.
- Validating the performance of both building and processing the tree at runtime.
